zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceCancelProcessor.java

```diff
@@ -135,17 +136,24 @@ private boolean validateCommand(
     return true;
   }
 
-  private long getRootProcessInstanceKey(final long instanceKey) {
+  private long getRootProcessInstanceKey(long instanceKey) {
+    var parentInstanceKey = getParentInstanceKey(instanceKey);
+    while (parentInstanceKey.isPresent()) {
+      instanceKey = parentInstanceKey.get();
+      parentInstanceKey = getParentInstanceKey(instanceKey);
+    }
+
+    return instanceKey;
+  }
 
+  private Optional<Long> getParentInstanceKey(final long instanceKey) {
     final var instance = elementInstanceState.getInstance(instanceKey);
     if (instance != null) {
-
       final var parentProcessInstanceKey = instance.getValue().getParentProcessInstanceKey();
       if (parentProcessInstanceKey > 0) {
-
-        return getRootProcessInstanceKey(parentProcessInstanceKey);
+        return Optional.of(parentProcessInstanceKey);
       }
     }
-    return instanceKey;
+    return Optional.empty();
```

Comment:
🔧 Personally I find this easier to read. But your solution is definitely not wrong, so I'll leave it up to you if you want to change it.

```java
  private long getRootProcessInstanceKey(long instanceKey) {
    var instance = elementInstanceState.getInstance(instanceKey);
    while (instance != null && instance.getValue().hasParentProcess()) {
      instanceKey = instance.getValue().getParentProcessInstanceKey();
      instance = elementInstanceState.getInstance(instanceKey);
    }
    
    return instanceKey;
  }
```